components/gantry/oneaxis/setup.go

```diff
@@ -0,0 +1,390 @@
+	// set up limits, range and knows absolute position along the axis
+	g, err := newOneAxis(ctx, deps, cfg, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	conf, ok := cfg.ConvertedAttributes.(*AttrConfig)
+	if !ok {
+		return nil, rdkutils.NewUnexpectedTypeError(conf, cfg.ConvertedAttributes)
+	}
+
+	oAx, ok := g.(*oneAxis)
+	if !ok {
+		return nil, rdkutils.NewUnexpectedTypeError(&oneAxis{}, oAx)
+	}
+
+	switch len(conf.LimitSwitchPins) {
+	case 0:
+		// homes the motor and finds limits using the motor's position reporting
+		if err = oAx.homeEncoder(ctx); err != nil {
+			return nil, err
+		}
+		return oAx, nil
+	default:
+		// sets up a gantry that requires polling of external sensors, in this case limit switches
+		// to find position limits and gantry range
+		g, err := newLimitSwitchGantry(ctx, cfg, logger, oAx)
+		if err != nil {
+			return nil, err
+		}
+
+		lAx, ok := g.(*limitSwitchGantry)
+		if !ok {
+			return nil, rdkutils.NewUnexpectedTypeError(&limitSwitchGantry{}, lAx)
+		}
+		if err = lAx.homeWithLimSwitch(ctx, conf.LimitSwitchPins); err != nil {
+			return nil, err
+		}
+		return g, nil
+	}
+}
+
+type limitSwitchGantry struct {
+	ctx context.Context
+	oAx *oneAxis
+
+	limitSwitchPins []string
+	limitHigh       bool
+	limitHitMu      sync.Mutex
+	limitsPolled    bool
+
+	mu     sync.Mutex
+	logger golog.Logger
+	generic.Unimplemented
+}
+
+// newOneAxis creates a new one axis gantry.
+func newLimitSwitchGantry(
+	ctx context.Context,
+	cfg config.Component,
+	logger golog.Logger,
+	oneAxis *oneAxis,
+) (gantry.LocalGantry, error) {
+	conf, ok := cfg.ConvertedAttributes.(*AttrConfig)
+	if !ok {
+		return nil, rdkutils.NewUnexpectedTypeError(conf, cfg.ConvertedAttributes)
+	}
+
+	if len(conf.LimitSwitchPins) > 2 {
+		return nil, errBadNumLimitSwitches(cfg.Name, len(conf.LimitSwitchPins))
+	}
+
+	g := &limitSwitchGantry{
+		ctx:             ctx,
+		limitSwitchPins: conf.LimitSwitchPins,
+		limitHigh:       *conf.LimitPinEnabled,
+		oAx:             oneAxis,
+		logger:          logger,
+		limitsPolled:    false,
+	}
+
+	if len(g.limitSwitchPins) == 1 && (g.oAx.mmPerRevolution == 0 || g.oAx.lengthMm == 0) {
+		return nil, errDimensionsNotFound(g.oAx.name, g.oAx.mmPerRevolution, g.oAx.mmPerRevolution)
+	}
+
+	g.limitHitChecker()
+
+	return g, nil
+}
+
+func (g *limitSwitchGantry) limitHitChecker() {
+	// lock the mutex associated with this monitoring function
+	g.limitHitMu.Lock()
+	// check if the state is false
+	if !g.limitsPolled {
+		return
+	}
+	// we know we are monitoring the limit switches in this gantry,
+	// return if the limitHits are already being checked
+	g.limitsPolled = true
+	g.limitHitMu.Unlock()
+
+	var errs error
+	for {
+		select {
+		case <-g.ctx.Done():
+			return
+		default:
+		}
+		for idx := range g.limitSwitchPins {
+			hit, err := g.limitHit(g.ctx, idx)
+			if hit || err != nil {
+				g.limitHitMu.Lock()
+				g.limitsPolled = false
+				g.limitHitMu.Unlock()
+				errs := multierr.Combine(errs, err)
```

Comment:
This line seems unnecessary: we know `errs` is nil because it's nil the first time through and every iteration except the first one, we would have returned at the end of the previous iteration if it was not nil.

So, I think the 4 lines starting here can all be condensed:
```go
// motor operation manager cancels running
errs = multierr.Combine(err, g.oAx.motor.Stop(g.ctx, nil))
```

Reply:
The `:=` operator creates a brand new variable `errs` which shadows the one created on line 217. I suspect line 217 should be removed, though I could be wrong about that.